src/reducers/pause.js

```diff
@@ -216,40 +272,63 @@ function update(
       // Preserving for the old debugger
       prefs.ignoreCaughtExceptions = !shouldPauseOnCaughtExceptions;
 
-      return {
-        ...state,
+      const newState = {
+        ...threadState(),
         shouldPauseOnExceptions,
         shouldPauseOnCaughtExceptions
       };
 
+      return updateThreadState(newState);
+    }
+
     case "COMMAND": {
-      return action.status === "start"
-        ? {
-            ...state,
-            ...emptyPauseState,
-            command: action.command,
-            previousLocation: getPauseLocation(state, action)
-          }
-        : { ...state, command: null };
+      const newState =
+        action.status === "start"
+          ? {
+              ...threadState(),
+              ...resumedPauseState,
+              command: action.command,
+              previousLocation: getPauseLocation(threadState(), action)
+            }
+          : { ...threadState(), command: null };
+
+      return updateThreadState(newState);
     }
 
     case "RESUME":
-      return { ...state, ...emptyPauseState };
+      // Workaround for threads resuming before the initial connection.
+      if (!action.thread && !state.currentThread) {
+        return state;
+      }
+      return updateThreadState({ ...threadState(), resumedPauseState });
 
-    case "EVALUATE_EXPRESSION":
-      return {
-        ...state,
+    case "EVALUATE_EXPRESSION": {
+      const newState = {
+        ...threadState(),
```

Comment:
it feels like we still have a lot of boilerplate
this feels better
```js
      return updateThreadState({
        ...threadState(),
        command: action.status === "start" ? "expression" : null
      });
```

... but i wonder if we could do a bit more

```js
updateThreadState({
   command: action.status === "start" ? "expression" : null
});
```

this means will always compute the threadState inside of update
```js
  const updateThreadState = threadState => {
    if (!action.thread) {
      throw new Error(`Missing thread in action ${action.type}`);
    }
    return {
      ...state,
      threads: {
        ...state.threads,
        [action.thread]: { ...threadState(), ...threadState }
      }
    };
  };
````